README.md
# Sony BRAVIA

Node.js module for discovering and controlling Sony BRAVIA Android TVs. This module allows you retrieve all the available service protocol API methods and invoke any of them. All methods return a [Promise](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Promise).

## Setup

### TV Setup

* Turn on your TV
* On the TV go to Settings > Network > Home network setup > Remote device/Renderer > On
* On the TV go to Settings > Network > Home network setup > IP Control > Authentication > Normal and Pre-Shared Key
* On the TV go to Settings > Network > Home network setup > Remote device/Renderer > Simple IP Control > On

**Optional:**

* On the TV go to Settings > Network > Home network setup > Remote device/Renderer > Enter Pre-Shared Key > 0000 (or whatever you want your PSK Key to be)

### Install with NPM

```sudo npm install @seydx/bravia@beta -g ```

## Discovery

```javascript
const Bravia = require('@seydx/bravia');
const bravia = new Bravia();

async function discover () {
  try {
    // Attempts to discover any Sony Bravia TVs.
    const devices = await bravia.discover();
    
    for (const device in devices) {
      console.log(devices[device]);
    }
  } catch (error) {
    console.error(error);    
  }
}
```

## Authentication

### Connect to TV via PSK

```javascript
// Connects to a Bravia TV at 192.168.1.2:80 with the PSK 0000.
const Bravia = require('@seydx/bravia');

const bravia = new Bravia({host: '192.168.1.2', port: 80, psk: '0000'});

```

### Connect to TV via PIN

To use the API with the PIN procedure, your credentials must first be created. Afterwards this credentials can be used to send further requests.


```javascript
// Connects to a Bravia TV at 192.168.1.2:80 and create your credentials.
const Bravia = require('@seydx/bravia');

const name = 'MyTV'; // Default: '@seydx/bravia'
const bravia = new Bravia({name: name, host: '192.168.1.2', port: 80, pin: true});

async function example(){
  try {
    const credentials = await bravia.pair();
    console.log(credentials)
  } catch(error) {
    console.log(error);
  }
}

```

The PIN displayed on the TV must then be entered in the terminal. This will generate a credentials ``<Object>`` like this:

```javascript
{
  name: 'MyTV',
  uuid: 'e9812807-d394-407c-b657-c89a98804e65',
  token: 'A0B9B9D7580466F22EE8F8EA148863774ACCE203',
  expires: 'Fr., 26 Apr. 2009 21:42:48 GMT+00:00'
}
```

With these credentials u can call the API without any authentication procedure

```javascript
// Connects to a Bravia TV at 192.168.1.2:80 with Application Name and UUID.
const Bravia = require('@seydx/bravia');

const credentials = {
  name: 'MyTV',
  uuid: 'e9812807-d394-407c-b657-c89a98804e65',
  token: 'A0B9B9D7580466F22EE8F8EA148863774ACCE203',
  expires: 'Fr., 26 Apr. 2009 21:42:48 GMT+00:00'
}

const bravia = new Bravia({host: '192.168.1.2', port: 80, ...credentials});
```

Alternatively, the credentials can also be created using the built-in CLI. See [#CLI](https://github.com/SeydX/bravia#command-line-tool)


## Usage

### Service Protocol APIs

```javascript
async function api(){
  try {
    // Retrieves all the system method types and versions.
    const methods = await bravia.describe();
    console.log(methods);
    
    // Retrieves all the available IRCC commands from the TV.
    const commands = await bravia.exec('system', 'getRemoteControllerInfo');
    console.log(commands);
    
    // Queries the volume info.
    const volume = await bravia.exec('audio', 'getVolumeInformation');
    console.log(volume);
    
    // Sets the speaker volume level to 50%.
    await bravia.exec('audio', 'setAudioVolume', '1.0', { target: 'speaker', volume: '50' });
  } catch(error) {
    console.log(error);
  }
}
```


### Send IRCC Code

```javascript
async function ircc(){
  try {
    // Retrieves all the available IRCC commands from the TV.
    const commands = await bravia.getIRCCCodes();
    
    // Sends an IRCC code signal by name.
    await bravia.execCommand('Mute');
    
    // Sends an IRCC code signal by value.
    await bravia.execCommand('AAAAAQAAAAEAAAAUAw==');
    
    // Sends multiple IRCC code signals by name and/or value. Change delay to alter time between each command sent.
    const delay = 350 //in milliseconds (Default: 350)
    
    await bravia.execCommand(['Hdmi1', 'AAAAAgAAABoAAABaAw==', 'Hdmi2', 'AAAAAgAAABoAAABbAw=='], delay);
  } catch(error) {
    console.log(error);
  }
}
```


### Turn on TV

The TV can easily be switched on via "Wake on LAN" or directly through the API. For WOL you need to enable WOL under TV settings.

Note: Only the screen from TVs in "PRO" mode can be switched ON with Wake-on-LAN. If your TV is in "NORMAL" mode and in standby, you need to enable the Rest API with WOL to perform API calls. 

More info: https://pro-bravia.sony.net/develop/integrate/ip-control/index.html#wake-on-lan

```javascript
async function turnOnTV(){
  try {
    // Optional (Default values)
    const options = {
      subnet: '255.255.255.255',
      num_packets: 10,
      interval: 100
    }

    // Turn on TV through Wake on LAN (WOL)
    await bravia.wake('33:7F:62:9F:7B:70', options)
    
    // Turn on TV through API
    await bravia.exec('system', 'setPowerStatus', '1.0', { status: true })
  } catch(error) {
    console.log(error);
  }
}
```

### Command line tool

The bravia cli support following commands:

- **pair** \<host> -p -n: Pair with a Bravia TV or refresh existing token (PIN Authentication)
  - \<host> (required): Bravia TV ip address.
  - -p, \--port (optional): Bravia TV port (Default: 80)
  - -n, \--name (optional): Custom name (Used for PIN Authentication, Default: '@seydx/bravia')
- **methods** \<host> -p: Retrieves all the system method types and versions
  - \<host> (required): Bravia TV ip address.
  - -p, \--port (optional): Bravia TV port (Default: 80)
- **exec** \<host> \<protocol> \<service> \<version> \<command> -p -n: Execute API call
  - \<host> (required): Bravia TV ip address.
  - \<protocol> (required): API Protocol (Endpoint)
  - \<service> (required): API Service
  - \<version> (optional): API Service Version (Default: '1.0')
  - \<command> (optional): API Command (Default: {})
  - -p, \--port (optional): Bravia TV port (Default: 80)
  - -n, \--name (optional): Custom name (Used for PIN Authentication, Default: '@seydx/bravia')
  - \--psk (optional): Pre-Shared Key (if not set, PIN authentication will be used)
- **discover**: Discover alls your TVs in network


#### Example usage for PIN Authentication

```
bravia pair 192.168.178.55 -p 80 -n MyTv
```

The PIN displayed on the TV must then be entered in the terminal. This will generate a credentials ``<Object>`` like this:

```javascript
name:  MyTv
uuid:  20879d92-1234-4ba3-a4ce-9a8444c71fa7
token: FD53E55779F964702178CDEBF71E3BA51A6D3A5D
expires: Fr., 26 Apr. 2009 21:42:48 GMT+00:00


CHANGELOG.md
# Changelog

# NEXT

## Notable Changes
- PIN authentication
- Checking available protocols, services and method types
- Improved CLI

## Other Changes
- The code has been refactored
- Other bug fixes
- Dependencies were updated

# v1.0.0
- Init (Fork https://github.com/waynehaffenden/bravia)

lib/request.js
'use-strict';

const debug = require('debug')('bravia:request');
const got = require('got');
const parseString = require('xml2js').parseStringPromise;

exports.request = async (uri, credentials, data, headers, pair) => {
  const options = {
    method: 'POST',
    headers: {
      'Content-Type': 'text/xml; charset=UTF-8',
      SOAPACTION: '"urn:schemas-sony-com:service:IRCC:1#X_SendIRCC"',
    },
  };

  if (data.xml) {
    options.body = data.xml;
    options.responseType = 'text';
  } else {
    options.json = data;
    options.responseType = 'json';
  }

  if (credentials.psk) {
    options.headers['X-Auth-PSK'] = credentials.psk;
  } else if (credentials.pin && credentials.pin.token) {
    options.headers.Cookie = `auth=${credentials.pin.token}`;
  }

  options.headers = {
    ...options.headers,
    ...headers,
  };

  debug(`Initializing request for ${uri}`);
  debug('Request %O', options);

  try {
    const response = await got(uri, options);
    response.body.turnedOff = null;

    //The “error” member must be an array and is defined as [error_code, error_message].
    //Error List: https://pro-bravia.sony.net/develop/integrate/rest-api/spec/errorcode-list/index.html
    const error = response.body.error;

    if (error) {
      if (error[1] === 'Illegal State') {
        response.body.result = [
          {
            uri: false,
            source: 'application',
            title: 'Application',
          },
        ];

        delete response.body.error;
      } else if (error[0] === 40005 || error[1] === 'Display Is Turned off' || error[1] === 'not power-on') {
        response.body.result = error;
        response.body.turnedOff = true;

        delete response.body.error;
      } else {
        const err = {
          response: {
            statusCode: error[0],
            statusMessage: error[1],
            body: '',
            url: response.url,
          },
        };

        throw err;
      }
    }

    //"response.body.result" must be an array type of fixed length. (The length is defined on each API specification.)
    return pair ? response : response.body;
  } catch (err) {
    if (err.response) {
      let error = new Error(`${err.response.statusCode} - ${err.response.statusMessage}`);
      let soapError = {};

      try {
        const result = await parseString(err.response.body);

        soapError = {
          errorCode: result['s:Envelope']['s:Body'][0]['s:Fault'][0].detail[0].UPnPError[0].errorCode[0],
          errorDescription: result['s:Envelope']['s:Body'][0]['s:Fault'][0].detail[0].UPnPError[0].errorDescription[0],
        };
      } catch {
        //unhandled
      }

      Object.assign(error, {
        title: 'Invalid Response',
        code: err.response.statusCode,
        message: err.response.statusMessage,
        soap: soapError,
        payload: {
          id: data.id,
          version: data.version,
          method: data.method,
          params: data.params && data.params.length ? data.params[0] : [],
        },
        url: err.response.url,
      });

      throw error;
    } else if (err.request) {
      let error = new Error(`${err.code} - ${err.message}`);

      Object.assign(error, {
        title: 'No Response',
        code: err.code,
        message: err.message,
        soap: {},
        payload: {
          id: data.id,
          version: data.version,
          method: data.method,
          params: data.params && data.params.length ? data.params[0] : [],
        },
        url: err.request.requestUrl,
      });

      throw error;
    } else {
      throw new Error(err);
    }
  }
};


lib/service.js
'use strict';

const debug = require('debug')('bravia:service');
const { request } = require('./request');

const paramsList = (params) => (Array.isArray(params) ? params : [params]);

class ServiceProtocol {
  constructor(protocol, url, credentials) {
    this.protocol = protocol;
    this.url = `${url}/${protocol}`;
    this.credentials = credentials;

    this.methods = [];
  }

  async getVersions() {
    const response = await this.invoke('getVersions');
    return response.result;
  }

  async describe(version) {
    if (this.methods.length > 0) {
      if (version) {
        return {
          protocol: this.protocol,
          methods: this.methods.filter((method) => method.version === version),
        };
      } else {
        return {
          protocol: this.protocol,
          methods: this.methods,
        };
      }
    }

    try {
      const versions = await this.getVersions();

      for (const version of versions) {
        if (version.length > 1) {
          for (const subversion of version) {
            const response = await this.invoke('getMethodTypes', '1.0', subversion);

            this.methods = this.methods.concat(
              response.results.map((method) => {
                return {
                  method: method[0],
                  version: method[method.length - 1],
                };
              })
            );
          }
        } else {
          const response = await this.invoke('getMethodTypes', '1.0', version);
          const results = response.results.map((method) => {
            return {
              method: method[0],
              version: method[method.length - 1],
            };
          });

          results.forEach((method) => this.methods.push(method));
        }
      }
    } catch (err) {
      if (err.code !== 404) {
        throw new Error(err);
      }
    }

    return {
      protocol: this.protocol,
      methods: this.methods,
    };
  }

  async invoke(method, version = '1.0', data = [], headers = {}) {
    /*
     * "id" is an integer and “0" must not be used. "0" is used as a special number system. The id can be set to "1"
     * Null valued parameters must be regarded as absent in the request/response, unless otherwise mentioned in each API specification
     * "params" must be an array type of fixed length.
     */

    const options = {
      id: 1,
      method: method,
      version: version,
      params: paramsList(data),
    };

    debug('Invoking /%s %O', this.protocol, options);

    return await request(this.url, this.credentials, options, headers);
  }
}

module.exports = ServiceProtocol;


src/service-protocol.js
'use strict';

class ServiceProtocol {
  constructor(bravia, protocol) {
    this.debug = require('debug')('bravia:service:' + protocol);

    this.bravia = bravia;
    this.protocol = protocol;
    this._methods = [];
  }

  async getVersions() {
    const versions = await this.invoke('getVersions');
    return versions;
  }

  async getMethodTypes(version) {
    if (this._methods.length > 0) {
      if (version) {
        let methods = this._methods.filter((method) => method.version === version);
        return methods;
      } else {
        return this._methods;
      }
    }

    const versions = await this.getVersions();

    for (const id of versions) {
      let data = await this.invoke('getMethodTypes', '1.0', id);
      let resolvedData = this.resolveMethods(data, id);

      this._methods.push({
        endpoint: this.protocol,
        version: id,
        availableMethods: resolvedData.availableMethods,
        availableParams: resolvedData.availableParams,
        methods: resolvedData.methods,
      });
    }

    return this._methods;
  }

  resolveMethods(data, id) {
    let availableMethods = [];
    let availableParams = {};

    let methods = data.map((method) => {
      this.debug('Registering method %s (%s) for %s', method[0], id, this.protocol);

      availableMethods.push(method[0]);
      availableParams[method[0]] = [];

      method[1].forEach((arg) => {
        if (arg.startsWith('{')) {
          if (arg.slice(-1) === '*') {
            Object.keys(JSON.parse(arg.substring(0, arg.length - 1))).forEach((arg2) => {
              this.debug('Registering parameter "%s" for %s (%s)', arg2, method[0], method[3]);
              availableParams[method[0]].push(arg2);
            });
          } else {
            Object.keys(JSON.parse(arg)).forEach((arg3) => {
              this.debug('Registering parameter "%s" for %s (%s)', arg3, method[0], method[3]);
              availableParams[method[0]].push(arg3);
            });
          }
        } else {
          this.debug('Registering parameter "%s" for %s (%s)', arg, method[0], method[3]);
          availableParams[method[0]].push(arg);
        }
      });

      return {
        method: method[0],
        in: method[1].map((arg) => {
          return arg.startsWith('{')
            ? arg.slice(-1) === '*'
              ? JSON.parse(arg.substring(0, arg.length - 1))
              : JSON.parse(arg)
            : arg;
        }),
        out: method[2].map((arg) => {
          return arg.startsWith('{')
            ? arg.slice(-1) === '*'
              ? JSON.parse(arg.substring(0, arg.length - 1))
              : JSON.parse(arg)
            : arg;
        }),
        version: method[3],
      };
    });

    return {
      availableMethods: availableMethods,
      availableParams: availableParams,
      methods: methods,
    };
  }

  // eslint-disable-next-line no-unused-vars
  async checkApiMethod(method, version, params) {
    if (!this.initialized) {
      this.debug('Method types not initialized yet. Initializing available method types before continue');
      await this.getMethodTypes();
      this.initialized = true;
    }

    let apiMethod = this._methods.filter((mthd) => mthd.availableMethods.includes(method));

    if (apiMethod.length) {
      let apiMethodByVersion = apiMethod.find((method) => method.version === version);

      if (apiMethodByVersion) {
        apiMethod = apiMethodByVersion;
      } else {
        apiMethod = apiMethod[apiMethod.length - 1];
      }

      if (apiMethod.version !== version)
        this.debug(
          'Requested version (%s) for %s could not be found. Replacing it with v%s',
          version,
          method,
          apiMethod.version
        );

      this.debug('Endpoint options %O', apiMethod);
      this.debug(
        'Method options %O',
        apiMethod.methods.find((availableMethod) => availableMethod.method === method)
      );

      return apiMethod;
    } else {
      this.debug('Requested method (' + method + ') could not be found!');

      return false;
    }
  }

  async invoke(method, version, params, turnOn) {
    let apiMethod;

    params = params ? [params] : [];
    version = version || '1.0';

    if (method !== 'getVersions' && method !== 'getMethodTypes') {
      apiMethod = await this.checkApiMethod(method, version, params);
      version = apiMethod ? apiMethod.version : version;
    }

    this.debug('Executing action %s (%s)', method, version);

    const response = await this.bravia._request({
      path: `/${this.protocol}`,
      json: {
        id: 3,
        method: method,
        version: version,
        params: params,
      },
      turnOn: turnOn,
    });

    if (response.data.results) {
      return response.data.results;
    } else if (response.data.result) {
      return response.data.result[response.data.result.length > 1 ? 1 : 0];
    }
  }
}

module.exports = ServiceProtocol;


lib/bravia.js
'use strict';

const debug = require('debug')('bravia:device');
const got = require('got');
const parseString = require('xml2js').parseStringPromise;
const prompt = require('enquirer').prompt;
const SsdpClient = require('node-ssdp').Client;
const URL = require('url').URL;
const wol = require('wake_on_lan');

const { request } = require('./request');
const Service = require('./service');
const SERVICE_PROTOCOLS = require('./protocols');
const utils = require('./utils');

const SSDP_SERVICE_TYPE = 'urn:schemas-sony-com:service:IRCC:1';
const DEFAULT_TIME_BETWEEN_COMMANDS = 350;

const DEFAULTS = {
  name: '@seydx/bravia',
  host: undefined,
  mac: undefined,
  port: 80,
  psk: false,
  pin: true,
};

const setTimeoutAsync = (ms) => new Promise((res) => setTimeout(res, ms));

class Bravia {
  constructor(options = {}) {
    this.initialized = false;

    this.options = {
      ...DEFAULTS,
      ...options,
      pin: options.psk ? false : true,
    };

    this.credentials = {};
    this.services = new Map();
    this.serviceMethods = [];
    this.irccCodes = [];

    if (this.options.psk) {
      this.credentials.psk = this.options.psk;
    }

    if (this.options.pin) {
      this.credentials.pin = {
        name: this.options.name,
        uuid: utils.genUUID(this.options.name),
        token: this.options.token,
        expires: this.options.expires,
      };
    }

    debug('Options set', this.options);
    debug('Credentials set', this.credentials);

    if (this.options.host) {
      this.url = new URL(`http://${this.options.host}:${this.options.port}/sony`);
      debug('Using url', this.url.toString());

      for (const protocol of SERVICE_PROTOCOLS) {
        debug(`Creating service /${protocol}`);
        this.services.set(protocol, new Service(protocol, this.url, this.credentials));
      }

      this.initialized = true;
    } else {
      debug('Module not initialized, no "host" defined!');
    }
  }

  async discover() {
    const ssdp = new SsdpClient();
    const discovered = [];

    debug('Starting discovery.');

    ssdp.search(SSDP_SERVICE_TYPE);
    ssdp.on('response', async (headers, statusCode, data) => {
      if (statusCode === 200) {
        try {
          const response = await got(headers.LOCATION);
          const result = await parseString(response.body);
          const device = result.root.device[0];

          if (device.serviceList) {
            // Not all devices return a serviceList (e.g. Philips Hue gateway responds without serviceList)
            const service = device.serviceList[0].service.find(
              (service) => service.serviceType[0] === SSDP_SERVICE_TYPE
            );

            const deviceUrl = new URL(service.controlURL[0].split('/ircc')[0]);

            if (!discovered.some((device) => device.host === deviceUrl.host)) {
              debug(`Found new device: ${device.friendlyName[0]} (${deviceUrl.host})`);

              const serviceMap = new Map();

              for (const protocol of SERVICE_PROTOCOLS) {
                debug(`Creating service /${protocol}`);
                serviceMap.set(protocol, new Service(protocol, deviceUrl, {}));
              }

              const discoveredSercices = Array.from(serviceMap.values());

              const services = await Promise.all(
                discoveredSercices.map(async (service) => {
                  const desc = await service.describe();
                  return {
                    service: desc.protocol,
                    methods: desc.methods,
                  };
                })
              );

              discovered.push({
                host: deviceUrl.host,
                port: deviceUrl.port || 80,
                friendlyName: device.friendlyName[0],
                manufacturer: device.manufacturer[0],
                manufacturerURL: device.manufacturerURL[0],
                modelName: device.modelName[0],
                UDN: device.UDN[0],
                services: services,
                scalar: device['av:X_ScalarWebAPI_DeviceInfo'],
                ircc: device['av:X_IRCC_DeviceInfo'],
                ircc_codes: device['av:X_IRCCCodeList'],
                rdis: device['av:X_RDIS_DeviceInfo'],
              });
            }
          }
        } catch (err) {
          ssdp.stop();
          throw new Error(`Error retrieving data for device ${data.address}. Error: ${err.message}`);
        }
      }
    });

    await setTimeoutAsync(5000);

    debug('Stopping discovery.');
    debug(`Found ${discovered.length} device(s)`);

    ssdp.stop();

    return discovered;
  }

  async pair(pin, refresh, manualAuth) {
    if (!this.initialized) {
      throw new Error('Module not initialized!');
    }

    const headers = {};

    if (pin) {
      debug(`Using PIN - ${pin} - for authentication`);

      const encodedPIN = Buffer.from(`:${pin}`, 'utf-8').toString('base64');
      headers.Authorization = `Basic ${encodedPIN}`;
    }

    const params = [
      {
        clientid: `${this.credentials.pin.name}:${this.credentials.pin.uuid}`,
        nickname: `${this.credentials.pin.name}`,
      },
      [
        {
          clientid: `${this.credentials.pin.name}:${this.credentials.pin.uuid}`,
          value: 'yes',
          nickname: `${this.credentials.pin.name}`,
          function: 'WOL',
        },
      ],
    ];

    try {
      const response = await request(
        `${this.url}/accessControl`,
        this.credentials,
        {
          id: 1,
          method: 'actRegister',
          version: '1.0',
          params: params,
        },
        headers,
        true
      );

      if (response.body.turnedOff) {
        throw new Error('Please turn on TV to handle authentication process through PIN.');
      }

      if (response.headers['set-cookie']) {
        this.credentials.pin = {
          ...this.credentials.pin,
          token: response.headers['set-cookie'][0].split(';')[0].split('auth=')[1],
          expires: response.headers['set-cookie'][0].split(';')[3].split('Expires=')[1],
        };

        debug('New token: %s', this.credentials.pin.token);

        return this.credentials.pin;
      } else {
        throw new Error(response);
      }
    } catch (err) {
      if (manualAuth) {
        throw err;
      }

      if (err.code === 401) {
        if (refresh) {
          throw new Error('Please use the CLI to register the TV before using this module!');
        }

        const pinPrompt = await prompt({
          type: 'input',
          name: 'pin',
          message: 'Please enter the four-digit PIN.',
        });

        return this.pair(pinPrompt.pin);
      } else {
        throw new Error(err);
      }
    }
  }

  async refreshToken() {
    if (this.credentials.pin) {
      const expireDateUnix = new Date(this.credentials.pin.expires).getTime();
      const nowDateUnix = new Date().getTime();

      if (!this.credentials.pin.token || expireDateUnix <= nowDateUnix) {
        await this.pair(false, true);
      }
    }

    return this.credentials;
  }

  async describe() {
    if (this.serviceMethods.length) {
      return this.serviceMethods;
    }

    const services = Array.from(this.services.values());
    this.serviceMethods = await Promise.all(
      services.map(async (service) => {
        const desc = await service.describe();
        return {
          service: desc.protocol,
          methods: desc.methods,
        };
      })
    );

    return this.serviceMethods;
  }

  async getIRCCCodes() {
    if (!this.initialized) {
      throw new Error('Module not initialized!');
    }

    if (this.irccCodes.length > 0) {
      return this.irccCodes;
    }

    const response = await this.exec('system', 'getRemoteControllerInfo');

    this.irccCodes = response.result[1];

    return this.irccCodes;
  }

  async exec(protocol, command, version = '1.0', data = [], headers = {}) {
    if (!this.initialized) {
      throw new Error('Module not initialized!');
    }

    const serviceMethods = await this.describe();
    const serviceProtocol = serviceMethods.find((serviceMethod) => serviceMethod.service === protocol);
    const serviceProtocolMap = this.services.get(protocol);

    if (!serviceProtocol || !serviceProtocolMap) {
      debug('Available services:', this.services.keys());

      let error = new Error(`Service Protocol "${protocol}" not known!`);

      Object.assign(error, {
        title: `Unknown Service Protocol: "${protocol}"`,
        code: 'APIERROR',
        message: `Service Protocol "${protocol}" not known!`,
        soap: {},
        payload: data,
        url: `${this.url}/${protocol}`,
      });

      throw error;
    } else {
      const method = serviceProtocol.methods.filter((serviceMethod) => serviceMethod.method === command);

      if (!method.length) {
        debug('Available methods:', serviceProtocol.methods);

        let error = new Error(`Service Method "${command}" not known!`);

        Object.assign(error, {
          title: `Unknown Service Method: "${command}"`,
          code: 'APIERROR',
          message: `Service Method "${command}" not known!`,
          soap: {},
          payload: data,
          url: `${this.url}/${protocol}`,
        });

        throw error;
      } else {
        const versionExist = method.find((availableMethod) => availableMethod.version === version);

        if (!versionExist) {
          const oldVersion = version;
          version = method[method.length - 1].version; //use always latest version

          debug(
            `Method version "${oldVersion}" does not exist for "${command}"! It will be replaced with latest available version: "${version}".`
          );
        }
      }
    }

    await this.refreshToken();

    return serviceProtocolMap.invoke(command, version, data, headers);
  }

  async execCommand(codes, delay) {
    if (!this.initialized) {
      throw new Error('Module not initialized!');
    }

    await this.refreshToken();
    await this.getIRCCCodes();

    delay = delay || DEFAULT_TIME_BETWEEN_COMMANDS;

    if (!Array.isArray(codes)) {
      codes = [codes];
    }

    for (const code of codes) {
      const irccCode = this.irccCodes.find((irccCode) => irccCode.name === code || irccCode.value === code);

      if (!irccCode) {
        let error = new Error(`IRCC code "${code}" not known!`);

        Object.assign(error, {
          title: `Unknown IRCC: "${code}"`,
          code: 'APIERROR',
          message: `IRCC code "${code}" not known!`,
          soap: {},
          payload: {
            code: codes,
          },
          url: `${this.url}/ircc`,
        });

        throw error;
      }

      debug(`Send ircc command: ${irccCode.name} (${irccCode.value})`);

      await request(`${this.url}/ircc`, this.credentials, {
        xml: `<?xml version="1.0"?>
        <s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/" s:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/">
            <s:Body>
                <u:X_SendIRCC xmlns:u="urn:schemas-sony-com:service:IRCC:1">
                    <IRCCCode>${irccCode.value}</IRCCCode>
                </u:X_SendIRCC>
            </s:Body>
        </s:Envelope>`,
      });

      if (codes.length > 1) {
        await setTimeoutAsync(delay);
      }
    }

    return;
  }

  async wake(mac, options = {}) {
    mac = mac || this.options.mac;

    if (/^([0-9A-F]{2}[:-]){5}([0-9A-F]{2})$/.test(mac)) {
      return new Promise((resolve, reject) => {
        mac = mac || this.options.mac;

        options = {
          address: options.subnet || '255.255.255.255',
          num_packets: options.num_packets || 15,
          interval: options.interval || 100,
          port: options.port || 9,
        };

        debug('Sending magic packets to %s', mac);
        debug('WOL Options %O', options);

        //https://pro-bravia.sony.net/develop/integrate/ip-control/index.html#wake-on-lan
        wol.wake(mac, options, async (error) => {
          if (error) {
            return reject(error);
          }

          try {
            debug('Waking up TV through API...');
            await this.exec('system', 'setPowerStatus', '1.0', { status: true });
          } catch (err) {
            reject(err);
          } finally {
            debug(`Magic packets send to ${mac}`);
            resolve();
          }
        });
      });
    } else {
      debug('No valid MAC address! Waking up TV through API...');
      await this.exec('system', 'setPowerStatus', '1.0', { status: true });
    }
  }
}

module.exports = Bravia;
